File: library-management/csharp/src/LibraryManagement/Reservation.cs
namespace LibraryManagement;

public class Reservation
{
    public const int ReservationExpiryDays = 3;

    public Reservation(Member member, Isbn isbn, DateOnly reservedOn)
    {
        Member = member;
        Isbn = isbn;
        ReservedOn = reservedOn;
    }

    public Member Member { get; }
    public Isbn Isbn { get; }
    public DateOnly ReservedOn { get; }
    public DateOnly? NotifiedOn { get; private set; }

    public bool IsNotified => NotifiedOn is not null;

    public bool HasExpiredAt(DateOnly today) =>
        NotifiedOn is DateOnly n && today.DayNumber - n.DayNumber > ReservationExpiryDays;

    internal void MarkNotified(DateOnly today) => NotifiedOn = today;
}

File: library-management/typescript/src/Reservation.ts
import { Isbn } from './Isbn.js';
import { Member } from './Member.js';

export const RESERVATION_EXPIRY_DAYS = 3;

const MS_PER_DAY = 86_400_000;

export class Reservation {
  private _notifiedOn: Date | null = null;

  constructor(
    public readonly member: Member,
    public readonly isbn: Isbn,
    public readonly reservedOn: Date,
  ) {}

  get notifiedOn(): Date | null { return this._notifiedOn; }
  get isNotified(): boolean { return this._notifiedOn !== null; }

  hasExpiredAt(today: Date): boolean {
    if (this._notifiedOn === null) return false;
    const daysSince = Math.round((today.getTime() - this._notifiedOn.getTime()) / MS_PER_DAY);
    return daysSince > RESERVATION_EXPIRY_DAYS;
  }

  markNotified(today: Date): void { this._notifiedOn = today; }
}
